### PR TITLE
[Core] Plugin name check

### PR DESCRIPTION
**Pull request by p-durandin**
### Details:
 -Check plugin name for loaded library
 - *...*

### Tickets:
 - *ticket-id*

### CODE REVIEW

File: src/inference/src/dev/core_impl.cpp [hunk 1]
```diff
@@ -592,6 +592,9 @@ ov::Plugin ov::CoreImpl::get_plugin(const std::string& pluginName) const {
             so = ov::util::load_shared_object(desc.libraryLocation.c_str());
             std::shared_ptr<ov::IPlugin> plugin_impl;
             reinterpret_cast<ov::CreatePluginFunc*>(ov::util::get_symbol(so, ov::create_plugin_function))(plugin_impl);
+            // Try to test plugin name
+            if(plugin_impl->get_device_name() != deviceName)
```
Comment (jane-intel) [inline src/inference/src/dev/core_impl.cpp:596]:
```suggestion
            if (plugin_impl->get_device_name() != deviceName)
```

**Review by olpipi — APPROVED** [review]

File: src/inference/src/dev/core_impl.cpp [hunk 2]
```diff
@@ -592,6 +592,9 @@ ov::Plugin ov::CoreImpl::get_plugin(const std::string& pluginName) const {
             so = ov::util::load_shared_object(desc.libraryLocation.c_str());
             std::shared_ptr<ov::IPlugin> plugin_impl;
             reinterpret_cast<ov::CreatePluginFunc*>(ov::util::get_symbol(so, ov::create_plugin_function))(plugin_impl);
+            // Try to test plugin name
+            if (plugin_impl->get_device_name() != deviceName)
+                OPENVINO_THROW("Device loading error");
```
Comment (praasz) [inline src/inference/src/dev/core_impl.cpp:597]:
```suggestion
            OPENVINO_ASSERT(plugin_impl->get_device_name() != deviceName, "Device loading error");
```

**Review by t-jankowski — APPROVED** [review]
Consider using `OPENVINO_ASSERT` as Paweł pointed out.

File: src/inference/src/dev/core_impl.cpp [hunk 3]
```diff
@@ -592,6 +592,14 @@ ov::Plugin ov::CoreImpl::get_plugin(const std::string& pluginName) const {
             so = ov::util::load_shared_object(desc.libraryLocation.c_str());
             std::shared_ptr<ov::IPlugin> plugin_impl;
             reinterpret_cast<ov::CreatePluginFunc*>(ov::util::get_symbol(so, ov::create_plugin_function))(plugin_impl);
+            auto plugin_name = plugin_impl->get_device_name();
+
+            // Check that device plugin name is the same as requested
+            // skip for meta
+            // TODO: Create API to detect meta plugins
+            if (plugin_name != "AUTO" && plugin_name != "MULTI") {
```
Comment (praasz) [inline src/inference/src/dev/core_impl.cpp:600]: Maybe this [function](https://github.com/openvinotoolkit/openvino/blob/a349dc82f9ac3bca06db3d052ecf37ab1de8b042/src/inference/src/dev/core_impl.cpp#L67) should be in dev API header to detect real/virtual device. This check can be not enough.

File: src/inference/src/dev/core_impl.cpp [hunk 4]
```diff
@@ -592,6 +592,14 @@ ov::Plugin ov::CoreImpl::get_plugin(const std::string& pluginName) const {
...
+                OPENVINO_ASSERT(plugin_name == deviceName, "Error: Device loaded with incorrect name\"", plugin_name, deviceName);
```
Comment (praasz) [inline src/inference/src/dev/core_impl.cpp:601]:
```suggestion
                OPENVINO_ASSERT(plugin_name == deviceName, deviceName, " device loaded with incorrect name: ", plugin_name);
```
?
Should give result as: `CPU device loaded with incorrect name: GPU`

**Comment by praasz** [comment]
Can be ticket number added to PR description (if there is any)?

File: src/inference/src/dev/core_impl.cpp [hunk 5]
```diff
@@ -592,6 +592,15 @@ ov::Plugin ov::CoreImpl::get_plugin(const std::string& pluginName) const {
             so = ov::util::load_shared_object(desc.libraryLocation.c_str());
             std::shared_ptr<ov::IPlugin> plugin_impl;
             reinterpret_cast<ov::CreatePluginFunc*>(ov::util::get_symbol(so, ov::create_plugin_function))(plugin_impl);
+            auto plugin_name = plugin_impl->get_device_name();
```
Comment (praasz) [inline src/inference/src/dev/core_impl.cpp:595]:
```suggestion
            const auto& plugin_name = plugin_impl->get_device_name();
```
Apply if there will be any change to not restart CI. But I think Coverity will complain about make a copy here.

**Review by praasz — APPROVED** [review]

File: src/inference/src/dev/core_impl.cpp [hunk 6]
```diff
@@ -592,6 +592,15 @@ ov::Plugin ov::CoreImpl::get_plugin(const std::string& pluginName) const {
...
             std::shared_ptr<ov::IPlugin> plugin_impl;
             reinterpret_cast<ov::CreatePluginFunc*>(ov::util::get_symbol(so, ov::create_plugin_function))(plugin_impl);
+            const auto& plugin_name = plugin_impl->get_device_name();
+
+            // Check that device plugin name is the same as requested for HW plugins
+            if (!plugin_name.empty() && !is_virtual_device(plugin_name)) {
+                OPENVINO_ASSERT(deviceName.find(plugin_name) != std::string::npos,
+                                deviceName,
+                                " device loaded with incorrect name: ",
+                                plugin_name);
```
Comment (ilya-lavrenov) [inline src/inference/src/dev/core_impl.cpp:602]: maybe we need to write a message that file "libraryLocation" is used for device "deviceName", while it contains implementation for "plugin_name" ?

**Review by ilya-lavrenov — APPROVED** [review]

**Pull request merged**